src/common/config_proxy.h

```diff
@@ -111,31 +184,39 @@ class ConfigProxy {
   }
   // for those want to reexpand special meta, e.g, $pid
   void finalize_reexpand_meta() {
-    std::lock_guard l(lock);
-    if (config.finalize_reexpand_meta(values, obs_mgr)) {
-      obs_mgr.apply_changes(values.changed, *this, nullptr);
-      values.changed.clear();
+    rev_obs_map_t rev_obs;
+    {
+      std::lock_guard l(lock);
+      if (config.finalize_reexpand_meta(values, obs_mgr)) {
+        _gather_changes(values.changed, &rev_obs, nullptr);
+        values.changed.clear();
+      }
     }
+
+    call_observers(rev_obs);
   }
   void add_observer(md_config_obs_t* obs) {
     std::lock_guard l(lock);
     obs_mgr.add_observer(obs);
+    obs_call_gate.emplace(obs, std::make_unique<CallGate>());
   }
   void remove_observer(md_config_obs_t* obs) {
     std::lock_guard l(lock);
+    call_gate_close(obs);
+    obs_call_gate.erase(obs);
```

Comment:
Sorry to revive this old PR, but I am hunting down a crash and I believe this isn't safe. If you remove an observer from this map between the time the copy of the obs list is made and before it can invoke `call_gate_leave`, `call_gate_leave` will fail to find the observer in the map and assert.

Reply:
... opened PR #29809